enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,246 @@
+---
+title: component-selection-during-install
+authors:
+  - "@bparees"
+reviewers:
+  - "@decarr"
+  - "@staebler"
+approvers:
+  - "@decarr"
+creation-date: 2021-05-04
+last-updated: 2021-05-04
+status: provisional
+---
+
+# User Selectable Install Solutions
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes a mechanism for cluster installers to exclude one or more optional components for
+their installation which will determine which payload components are/are not installed in their cluster.
+Core components are defined as the set of Second Level Operators managed by the Cluster Version Operator
+which today cannot be disabled until after completing the install and editing a CVO override, or editing
+the CVO overrides as part of rendering+editing manifests.
+
+The proposed UX is to make this a first class part of the install config api with the implementation
+being arguments supplied to the CVO to filter the user-selected manifests.
```

Comment:
I think there is more involved in the installer implementation of this than simply passing along arguments to the CVO.
1. Presumably the installer would need to know which addons can be excluded so that the installer can do validation on the user input.
2. Some addons may require the installer to omit manifests that it would normally create. For example, if the baremetal-operator addons were omitted, then the installer would need to omit the [manifest that configures the metal provisioning](https://github.com/openshift/installer/blob/master/data/data/manifests/openshift/baremetal-provisioning-config.yaml.template).
3. There would probably be an expectation by the user that the installer would make some opinionated decisions on behalf of the user about which addons to exclude. Again with the baremetal-operator addon, my expectation as a user would be that the installer would configure the exclusions automatically so that the baremetal-operator addon was excluded when using a platform other than "baremetal".

Reply:
1) maybe?  specifying things to exclude that don't actually exclude anything isn't a great UX but i dunno that it's a requirement that we tell you your exclusion didn't match anything.

2) not sure i follow this one.  I'd expect all related manifests to be annotated such that they get excluded as a group.  Could the manifest that configures the metal provisioning not also be annotated?   Is it instantiated by the CVO or by the installer itself?

3) i'd classify that as a potential future goal/possibility, but not a requirement here

Reply:
on the same topic - will the user be able to say exclude bm-operator but install bm platform? If not, how the installer will know to block it?

Reply:
I tried to cover these sort of scenarios here:
https://github.com/openshift/enhancements/pull/784/files#diff-4b412132f6f354bd03403cc85e104b445d5ff581abebae3d0b6fb69c58f094e7R120-R122

so we can discuss if that's "sufficient/acceptable" or not, but the position of this EP was to punt on such things.

Reply:
`maybe? specifying things to exclude that don't actually exclude anything isn't a great UX but i dunno that it's a requirement that we tell you your exclusion didn't match anything.
`
I think it will be a bad experience if the user will have a typo. Maybe, we can find a way for the CVO to verify it during bootstrap and fail early. kinda cvo-render --verify-components

Reply:
>     2. not sure i follow this one.  I'd expect all related manifests to be annotated such that they get excluded as a group.  Could the manifest that configures the metal provisioning not also be annotated?   Is it instantiated by the CVO or by the installer itself?

It is ultimately added to the cluster via CVO. So the manifest could have the annotation. The installer would need to add that annotation since the installer creates the manifest.

Reply:
> on the same topic - will the user be able to say exclude bm-operator but install bm platform? If not, how the installer will know to block it?

Seems like that is the scope of this proposal, but as @staebler mentions it implies some additional validation in the installer, e.g taking the baremetal example - it's reasonable to do `create manifests` with the baremetal SLO disabled (perhaps to deploy via UPI or assisted where those components may not be required), but a user should not be allowed to do `create cluster` with that configuration, because deployment of any worker hosts will fail.

Reply:
We have more recent requirements to be able to install the metal tools on platforms other than `baremetal`, so that's probably no longer a good example of something we would expect the installer to exclude automatically.

That said, I think we can ignore the automated guidance angle, at least in a first iteration. We can make it fancier/smarter later, but the first version should have simple rules: The default guidance is to install everything. If you want less than everything, you tell the CVO and it leaves out the things you don't want.